packages/smooth_app/lib/pages/smooth_bottom_navigation_bar.dart

```diff
@@ -1,85 +1,109 @@
-        showUnselectedLabels: false,
-        selectedItemColor: Colors.white,
-        backgroundColor: Theme.of(context).appBarTheme.backgroundColor,
-        currentIndex: _tabs.indexOf(tab),
-        onTap: (final int index) async => Navigator.push<Widget>(
-          context,
-          MaterialPageRoute<Widget>(
-            builder: (BuildContext context) => _getTabPage(_tabs[index]),
-          ),
+  Widget build(BuildContext context) {
+    return WillPopScope(
+      onWillPop: () async {
+        final bool isFirstRouteInCurrentTab =
+            !await _navigatorKeys[_currentPage]!.currentState!.maybePop();
+        if (isFirstRouteInCurrentTab) {
+          if (_currentPage != SmoothBottomNavigationTab.Scan) {
+            _selectTab(SmoothBottomNavigationTab.Scan, 1);
+            return false;
+          }
+        }
+        // let system handle back button if we're on the first route
+        return isFirstRouteInCurrentTab;
+      },
+      child: Scaffold(
+        body: Stack(children: <Widget>[
+          _buildOffstageNavigator(SmoothBottomNavigationTab.Profile),
+          _buildOffstageNavigator(SmoothBottomNavigationTab.Scan),
+          _buildOffstageNavigator(SmoothBottomNavigationTab.History),
+        ]),
+        bottomNavigationBar: BottomNavigationBar(
+          showSelectedLabels: false,
+          showUnselectedLabels: false,
+          selectedItemColor: Colors.white,
+          backgroundColor: Theme.of(context).appBarTheme.backgroundColor,
+          onTap: (int index) {
+            _selectTab(pageKeys[index], index);
+          },
+          currentIndex: _selectedIndex,
+          items: const <BottomNavigationBarItem>[
+            BottomNavigationBarItem(
+              icon: Icon(Icons.account_circle),
+              label: 'Profile',
+            ),
+            BottomNavigationBarItem(
+              icon: Icon(Icons.search),
+              label: 'Scan or Search',
+            ),
+            BottomNavigationBarItem(
+              icon: Icon(Icons.history),
+              label: 'History',
+            ),
+          ],
         ),
-        items: <BottomNavigationBarItem>[
-          _buildItem(_pages[_tabs[0]]!),
-          _buildItem(_pages[_tabs[1]]!),
-          _buildItem(_pages[_tabs[2]]!),
-        ],
-      );
+      ),
+    );
+  }
 
-  BottomNavigationBarItem _buildItem(final _Page page) =>
-      BottomNavigationBarItem(
-        icon: Icon(page.icon, size: 28),
-        label: page.name,
-      );
+  Widget _buildOffstageNavigator(SmoothBottomNavigationTab tabItem) {
+    final bool offstage = _currentPage != tabItem;
+    return Offstage(
+      offstage: offstage,
+      child: tabItem != SmoothBottomNavigationTab.Scan
+          ? TabNavigator(
+              navigatorKey: _navigatorKeys[tabItem]!,
+              tabItem: tabItem,
+            )
+          : ScanPage(
+              offstage: offstage,
+              navigatorKey: _navigatorKeys[tabItem]!,
+            ),
```

Comment:
This is confusing: why would the scan page be considered differently?

Reply:
Added a comment for that